unstruct/interface.go

```diff
@@ -0,0 +1,96 @@
+package unstruct
+
+// Map lets you manipulate an object.
+type Map interface {
+	// Data returns the underlying data provided by the object.
+	Data() map[string]interface{}
+
+	// Parent returns the parent value or nil if no parent is
+	// present (root).
+	Parent() Value
+
+	// Field returns the Value object in the "key" field.
+	//
+	// If this key is not present in the map, this function will
+	// create, set and return a new Value for this field, with a nil
+	// content.
+	Field(key string) Value
+
+	// HasField returns true if the field exists.
+	HasField(key string) bool
+
+	// Fields returns the sorted list of fields in the map.
+	Fields() []string
+}
+
+// Slice lets you manipulate an array.
+type Slice interface {
+	// Data returns the underlying data provided by the object.
+	Data() []interface{}
+
+	// Parent returns the parent value or nil if no parent is
+	// present (root).
+	Parent() Value
+
+	// Length returns the number of items in the slice.
+	Length() int
+
+	// At returns the "index"-th item in the slice.
+	//
+	// If the index is out-of-bound, nil is returned.
+	At(index int) Value
+
+	// Append adds at the end of the slice if the current object is
+	// a slice.
+	//
+	// Returns a Value pointing to that item.
+	Append(value interface{}) Value
+
+	// InsertAt inserts at the index-th position in the slice.
+	//
+	// Returns the slice itself, so that the calls can be chained,
+	// or nil if the index is out-of-bound.
+	InsertAt(index int, value interface{}) Slice
```

Comment:
Insert implies preserving the subsequent contents and shifting them one.  Is that what this is really doing?

Reply:
Yes, should this be clarified in the comment? (but insert implies ...)